apps/web/src/pages/tenants/components/UpdateTenantSidebar.tsx

```diff
@@ -0,0 +1,137 @@
+import { Group, Stack } from '@mantine/core';
+import { Controller, useForm } from 'react-hook-form';
+import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
+import { format } from 'date-fns';
+
+import { ITenantEntity, IUpdateTenantBodyDto } from '@novu/shared';
+
+import { Button, colors, NameInput, Sidebar, Text } from '../../../design-system';
+import { getTenantByIdentifier, updateTenant } from '../../../api/tenants';
+import { errorMessage, successMessage } from '../../../utils/notifications';
+import { QueryKeys } from '../../../api/query.keys';
+import { useEffectOnce } from '../../../hooks';
+import { TenantFormCommonFields } from './TenantFormCommonFields';
+
+export interface ITenantForm {
+  identifier: string;
+  name: string;
+  data: string;
+}
+export const defaultFormValues = {
+  identifier: '',
+  name: '',
+  data: '{}',
+};
+export function UpdateTenantSidebar({
+  isOpened,
+  tenantIdentifier,
+  onClose,
+}: {
+  isOpened: boolean;
+  tenantIdentifier?: string;
+  onClose: () => void;
+}) {
+  const queryClient = useQueryClient();
+
+  const { data: tenant, isLoading: isLoadingTenant } = useQuery<ITenantEntity>(
+    [QueryKeys.getTenantByIdentifier(tenantIdentifier)],
+    () => getTenantByIdentifier(tenantIdentifier as string),
+    {
+      enabled: !!tenantIdentifier,
+    }
+  );
+
+  const { mutateAsync: updateTenantMutate, isLoading: isLoadingUpdate } = useMutation<
+    ITenantEntity,
+    { error: string; message: string; statusCode: number },
+    { identifier: string; data: IUpdateTenantBodyDto }
+  >(({ identifier, data }) => updateTenant(identifier, data), {
+    onSuccess: async () => {
+      await queryClient.refetchQueries({
+        predicate: ({ queryKey }) => queryKey.includes(QueryKeys.tenantsList),
+      });
+      successMessage('Tenant has been updated!');
+    },
+    onError: (e: any) => {
+      errorMessage(e.message || 'Unexpected error');
+    },
+  });
+
+  const { handleSubmit, control, formState, reset } = useForm<ITenantForm>({
+    shouldUseNativeValidation: false,
+    defaultValues: defaultFormValues,
+  });
+
+  useEffectOnce(() => {
+    if (!tenant) return;
+
+    reset({ name: tenant.name, identifier: tenant.identifier, data: JSON.stringify(tenant.data) });
+  }, !!tenant);
```

Comment:
We have here a potential race condition, when the user updated the data, and get to the editor but the first copy of the data is stale, then will be refetched but because we have `useEffectOnce` we won't update the data.

```suggestion
  useEffect(() => {
    if (!tenant) return;

    reset({ name: tenant.name, identifier: tenant.identifier, data: JSON.stringify(tenant.data) });
  }, [tenant]);
```

https://www.loom.com/share/eb152f5e8bb14aa69123dd40a3ade623

Reply:
Good catch. You are right. Thanks!